<!DOCTYPE html>
<html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8"/>

<title>Introduction</title>

<script type="text/javascript">
window.onload = function() {
  var imgs = document.getElementsByTagName('img'), i, img;
  for (i = 0; i < imgs.length; i++) {
    img = imgs[i];
    // center an image if it is the only element of its parent
    if (img.parentElement.childElementCount === 1)
      img.parentElement.style.textAlign = 'center';
  }
};
</script>

<!-- Styles for R syntax highlighter -->
<style type="text/css">
   pre .operator,
   pre .paren {
     color: rgb(104, 118, 135)
   }

   pre .literal {
     color: #990073
   }

   pre .number {
     color: #099;
   }

   pre .comment {
     color: #998;
     font-style: italic
   }

   pre .keyword {
     color: #900;
     font-weight: bold
   }

   pre .identifier {
     color: rgb(0, 0, 0);
   }

   pre .string {
     color: #d14;
   }
</style>

<!-- R syntax highlighter -->
<script type="text/javascript">
var hljs=new function(){function m(p){return p.replace(/&/gm,"&amp;").replace(/</gm,"&lt;")}function f(r,q,p){return RegExp(q,"m"+(r.cI?"i":"")+(p?"g":""))}function b(r){for(var p=0;p<r.childNodes.length;p++){var q=r.childNodes[p];if(q.nodeName=="CODE"){return q}if(!(q.nodeType==3&&q.nodeValue.match(/\s+/))){break}}}function h(t,s){var p="";for(var r=0;r<t.childNodes.length;r++){if(t.childNodes[r].nodeType==3){var q=t.childNodes[r].nodeValue;if(s){q=q.replace(/\n/g,"")}p+=q}else{if(t.childNodes[r].nodeName=="BR"){p+="\n"}else{p+=h(t.childNodes[r])}}}if(/MSIE [678]/.test(navigator.userAgent)){p=p.replace(/\r/g,"\n")}return p}function a(s){var r=s.className.split(/\s+/);r=r.concat(s.parentNode.className.split(/\s+/));for(var q=0;q<r.length;q++){var p=r[q].replace(/^language-/,"");if(e[p]){return p}}}function c(q){var p=[];(function(s,t){for(var r=0;r<s.childNodes.length;r++){if(s.childNodes[r].nodeType==3){t+=s.childNodes[r].nodeValue.length}else{if(s.childNodes[r].nodeName=="BR"){t+=1}else{if(s.childNodes[r].nodeType==1){p.push({event:"start",offset:t,node:s.childNodes[r]});t=arguments.callee(s.childNodes[r],t);p.push({event:"stop",offset:t,node:s.childNodes[r]})}}}}return t})(q,0);return p}function k(y,w,x){var q=0;var z="";var s=[];function u(){if(y.length&&w.length){if(y[0].offset!=w[0].offset){return(y[0].offset<w[0].offset)?y:w}else{return w[0].event=="start"?y:w}}else{return y.length?y:w}}function t(D){var A="<"+D.nodeName.toLowerCase();for(var B=0;B<D.attributes.length;B++){var C=D.attributes[B];A+=" "+C.nodeName.toLowerCase();if(C.value!==undefined&&C.value!==false&&C.value!==null){A+='="'+m(C.value)+'"'}}return A+">"}while(y.length||w.length){var v=u().splice(0,1)[0];z+=m(x.substr(q,v.offset-q));q=v.offset;if(v.event=="start"){z+=t(v.node);s.push(v.node)}else{if(v.event=="stop"){var p,r=s.length;do{r--;p=s[r];z+=("</"+p.nodeName.toLowerCase()+">")}while(p!=v.node);s.splice(r,1);while(r<s.length){z+=t(s[r]);r++}}}}return z+m(x.substr(q))}function j(){function q(x,y,v){if(x.compiled){return}var u;var s=[];if(x.k){x.lR=f(y,x.l||hljs.IR,true);for(var w in x.k){if(!x.k.hasOwnProperty(w)){continue}if(x.k[w] instanceof Object){u=x.k[w]}else{u=x.k;w="keyword"}for(var r in u){if(!u.hasOwnProperty(r)){continue}x.k[r]=[w,u[r]];s.push(r)}}}if(!v){if(x.bWK){x.b="\\b("+s.join("|")+")\\s"}x.bR=f(y,x.b?x.b:"\\B|\\b");if(!x.e&&!x.eW){x.e="\\B|\\b"}if(x.e){x.eR=f(y,x.e)}}if(x.i){x.iR=f(y,x.i)}if(x.r===undefined){x.r=1}if(!x.c){x.c=[]}x.compiled=true;for(var t=0;t<x.c.length;t++){if(x.c[t]=="self"){x.c[t]=x}q(x.c[t],y,false)}if(x.starts){q(x.starts,y,false)}}for(var p in e){if(!e.hasOwnProperty(p)){continue}q(e[p].dM,e[p],true)}}function d(B,C){if(!j.called){j();j.called=true}function q(r,M){for(var L=0;L<M.c.length;L++){if((M.c[L].bR.exec(r)||[null])[0]==r){return M.c[L]}}}function v(L,r){if(D[L].e&&D[L].eR.test(r)){return 1}if(D[L].eW){var M=v(L-1,r);return M?M+1:0}return 0}function w(r,L){return L.i&&L.iR.test(r)}function K(N,O){var M=[];for(var L=0;L<N.c.length;L++){M.push(N.c[L].b)}var r=D.length-1;do{if(D[r].e){M.push(D[r].e)}r--}while(D[r+1].eW);if(N.i){M.push(N.i)}return f(O,M.join("|"),true)}function p(M,L){var N=D[D.length-1];if(!N.t){N.t=K(N,E)}N.t.lastIndex=L;var r=N.t.exec(M);return r?[M.substr(L,r.index-L),r[0],false]:[M.substr(L),"",true]}function z(N,r){var L=E.cI?r[0].toLowerCase():r[0];var M=N.k[L];if(M&&M instanceof Array){return M}return false}function F(L,P){L=m(L);if(!P.k){return L}var r="";var O=0;P.lR.lastIndex=0;var M=P.lR.exec(L);while(M){r+=L.substr(O,M.index-O);var N=z(P,M);if(N){x+=N[1];r+='<span class="'+N[0]+'">'+M[0]+"</span>"}else{r+=M[0]}O=P.lR.lastIndex;M=P.lR.exec(L)}return r+L.substr(O,L.length-O)}function J(L,M){if(M.sL&&e[M.sL]){var r=d(M.sL,L);x+=r.keyword_count;return r.value}else{return F(L,M)}}function I(M,r){var L=M.cN?'<span class="'+M.cN+'">':"";if(M.rB){y+=L;M.buffer=""}else{if(M.eB){y+=m(r)+L;M.buffer=""}else{y+=L;M.buffer=r}}D.push(M);A+=M.r}function G(N,M,Q){var R=D[D.length-1];if(Q){y+=J(R.buffer+N,R);return false}var P=q(M,R);if(P){y+=J(R.buffer+N,R);I(P,M);return P.rB}var L=v(D.length-1,M);if(L){var O=R.cN?"</span>":"";if(R.rE){y+=J(R.buffer+N,R)+O}else{if(R.eE){y+=J(R.buffer+N,R)+O+m(M)}else{y+=J(R.buffer+N+M,R)+O}}while(L>1){O=D[D.length-2].cN?"</span>":"";y+=O;L--;D.length--}var r=D[D.length-1];D.length--;D[D.length-1].buffer="";if(r.starts){I(r.starts,"")}return R.rE}if(w(M,R)){throw"Illegal"}}var E=e[B];var D=[E.dM];var A=0;var x=0;var y="";try{var s,u=0;E.dM.buffer="";do{s=p(C,u);var t=G(s[0],s[1],s[2]);u+=s[0].length;if(!t){u+=s[1].length}}while(!s[2]);if(D.length>1){throw"Illegal"}return{r:A,keyword_count:x,value:y}}catch(H){if(H=="Illegal"){return{r:0,keyword_count:0,value:m(C)}}else{throw H}}}function g(t){var p={keyword_count:0,r:0,value:m(t)};var r=p;for(var q in e){if(!e.hasOwnProperty(q)){continue}var s=d(q,t);s.language=q;if(s.keyword_count+s.r>r.keyword_count+r.r){r=s}if(s.keyword_count+s.r>p.keyword_count+p.r){r=p;p=s}}if(r.language){p.second_best=r}return p}function i(r,q,p){if(q){r=r.replace(/^((<[^>]+>|\t)+)/gm,function(t,w,v,u){return w.replace(/\t/g,q)})}if(p){r=r.replace(/\n/g,"<br>")}return r}function n(t,w,r){var x=h(t,r);var v=a(t);var y,s;if(v){y=d(v,x)}else{return}var q=c(t);if(q.length){s=document.createElement("pre");s.innerHTML=y.value;y.value=k(q,c(s),x)}y.value=i(y.value,w,r);var u=t.className;if(!u.match("(\\s|^)(language-)?"+v+"(\\s|$)")){u=u?(u+" "+v):v}if(/MSIE [678]/.test(navigator.userAgent)&&t.tagName=="CODE"&&t.parentNode.tagName=="PRE"){s=t.parentNode;var p=document.createElement("div");p.innerHTML="<pre><code>"+y.value+"</code></pre>";t=p.firstChild.firstChild;p.firstChild.cN=s.cN;s.parentNode.replaceChild(p.firstChild,s)}else{t.innerHTML=y.value}t.className=u;t.result={language:v,kw:y.keyword_count,re:y.r};if(y.second_best){t.second_best={language:y.second_best.language,kw:y.second_best.keyword_count,re:y.second_best.r}}}function o(){if(o.called){return}o.called=true;var r=document.getElementsByTagName("pre");for(var p=0;p<r.length;p++){var q=b(r[p]);if(q){n(q,hljs.tabReplace)}}}function l(){if(window.addEventListener){window.addEventListener("DOMContentLoaded",o,false);window.addEventListener("load",o,false)}else{if(window.attachEvent){window.attachEvent("onload",o)}else{window.onload=o}}}var e={};this.LANGUAGES=e;this.highlight=d;this.highlightAuto=g;this.fixMarkup=i;this.highlightBlock=n;this.initHighlighting=o;this.initHighlightingOnLoad=l;this.IR="[a-zA-Z][a-zA-Z0-9_]*";this.UIR="[a-zA-Z_][a-zA-Z0-9_]*";this.NR="\\b\\d+(\\.\\d+)?";this.CNR="\\b(0[xX][a-fA-F0-9]+|(\\d+(\\.\\d*)?|\\.\\d+)([eE][-+]?\\d+)?)";this.BNR="\\b(0b[01]+)";this.RSR="!|!=|!==|%|%=|&|&&|&=|\\*|\\*=|\\+|\\+=|,|\\.|-|-=|/|/=|:|;|<|<<|<<=|<=|=|==|===|>|>=|>>|>>=|>>>|>>>=|\\?|\\[|\\{|\\(|\\^|\\^=|\\||\\|=|\\|\\||~";this.ER="(?![\\s\\S])";this.BE={b:"\\\\.",r:0};this.ASM={cN:"string",b:"'",e:"'",i:"\\n",c:[this.BE],r:0};this.QSM={cN:"string",b:'"',e:'"',i:"\\n",c:[this.BE],r:0};this.CLCM={cN:"comment",b:"//",e:"$"};this.CBLCLM={cN:"comment",b:"/\\*",e:"\\*/"};this.HCM={cN:"comment",b:"#",e:"$"};this.NM={cN:"number",b:this.NR,r:0};this.CNM={cN:"number",b:this.CNR,r:0};this.BNM={cN:"number",b:this.BNR,r:0};this.inherit=function(r,s){var p={};for(var q in r){p[q]=r[q]}if(s){for(var q in s){p[q]=s[q]}}return p}}();hljs.LANGUAGES.cpp=function(){var a={keyword:{"false":1,"int":1,"float":1,"while":1,"private":1,"char":1,"catch":1,"export":1,virtual:1,operator:2,sizeof:2,dynamic_cast:2,typedef:2,const_cast:2,"const":1,struct:1,"for":1,static_cast:2,union:1,namespace:1,unsigned:1,"long":1,"throw":1,"volatile":2,"static":1,"protected":1,bool:1,template:1,mutable:1,"if":1,"public":1,friend:2,"do":1,"return":1,"goto":1,auto:1,"void":2,"enum":1,"else":1,"break":1,"new":1,extern:1,using:1,"true":1,"class":1,asm:1,"case":1,typeid:1,"short":1,reinterpret_cast:2,"default":1,"double":1,register:1,explicit:1,signed:1,typename:1,"try":1,"this":1,"switch":1,"continue":1,wchar_t:1,inline:1,"delete":1,alignof:1,char16_t:1,char32_t:1,constexpr:1,decltype:1,noexcept:1,nullptr:1,static_assert:1,thread_local:1,restrict:1,_Bool:1,complex:1},built_in:{std:1,string:1,cin:1,cout:1,cerr:1,clog:1,stringstream:1,istringstream:1,ostringstream:1,auto_ptr:1,deque:1,list:1,queue:1,stack:1,vector:1,map:1,set:1,bitset:1,multiset:1,multimap:1,unordered_set:1,unordered_map:1,unordered_multiset:1,unordered_multimap:1,array:1,shared_ptr:1}};return{dM:{k:a,i:"</",c:[hljs.CLCM,hljs.CBLCLM,hljs.QSM,{cN:"string",b:"'\\\\?.",e:"'",i:"."},{cN:"number",b:"\\b(\\d+(\\.\\d*)?|\\.\\d+)(u|U|l|L|ul|UL|f|F)"},hljs.CNM,{cN:"preprocessor",b:"#",e:"$"},{cN:"stl_container",b:"\\b(deque|list|queue|stack|vector|map|set|bitset|multiset|multimap|unordered_map|unordered_set|unordered_multiset|unordered_multimap|array)\\s*<",e:">",k:a,r:10,c:["self"]}]}}}();hljs.LANGUAGES.r={dM:{c:[hljs.HCM,{cN:"number",b:"\\b0[xX][0-9a-fA-F]+[Li]?\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"number",b:"\\b\\d+(?:[eE][+\\-]?\\d*)?L\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"number",b:"\\b\\d+\\.(?!\\d)(?:i\\b)?",e:hljs.IMMEDIATE_RE,r:1},{cN:"number",b:"\\b\\d+(?:\\.\\d*)?(?:[eE][+\\-]?\\d*)?i?\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"number",b:"\\.\\d+(?:[eE][+\\-]?\\d*)?i?\\b",e:hljs.IMMEDIATE_RE,r:1},{cN:"keyword",b:"(?:tryCatch|library|setGeneric|setGroupGeneric)\\b",e:hljs.IMMEDIATE_RE,r:10},{cN:"keyword",b:"\\.\\.\\.",e:hljs.IMMEDIATE_RE,r:10},{cN:"keyword",b:"\\.\\.\\d+(?![\\w.])",e:hljs.IMMEDIATE_RE,r:10},{cN:"keyword",b:"\\b(?:function)",e:hljs.IMMEDIATE_RE,r:2},{cN:"keyword",b:"(?:if|in|break|next|repeat|else|for|return|switch|while|try|stop|warning|require|attach|detach|source|setMethod|setClass)\\b",e:hljs.IMMEDIATE_RE,r:1},{cN:"literal",b:"(?:NA|NA_integer_|NA_real_|NA_character_|NA_complex_)\\b",e:hljs.IMMEDIATE_RE,r:10},{cN:"literal",b:"(?:NULL|TRUE|FALSE|T|F|Inf|NaN)\\b",e:hljs.IMMEDIATE_RE,r:1},{cN:"identifier",b:"[a-zA-Z.][a-zA-Z0-9._]*\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"operator",b:"<\\-(?!\\s*\\d)",e:hljs.IMMEDIATE_RE,r:2},{cN:"operator",b:"\\->|<\\-",e:hljs.IMMEDIATE_RE,r:1},{cN:"operator",b:"%%|~",e:hljs.IMMEDIATE_RE},{cN:"operator",b:">=|<=|==|!=|\\|\\||&&|=|\\+|\\-|\\*|/|\\^|>|<|!|&|\\||\\$|:",e:hljs.IMMEDIATE_RE,r:0},{cN:"operator",b:"%",e:"%",i:"\\n",r:1},{cN:"identifier",b:"`",e:"`",r:0},{cN:"string",b:'"',e:'"',c:[hljs.BE],r:0},{cN:"string",b:"'",e:"'",c:[hljs.BE],r:0},{cN:"paren",b:"[[({\\])}]",e:hljs.IMMEDIATE_RE,r:0}]}};
hljs.initHighlightingOnLoad();
</script>

<!-- MathJax scripts -->
<script type="text/javascript" src="https://cdn.bootcss.com/mathjax/2.7.0/MathJax.js?config=TeX-MML-AM_CHTML">
</script>


<style type="text/css">
body, td {
   font-family: sans-serif;
   background-color: white;
   font-size: 13px;
}

body {
  max-width: 800px;
  margin: auto;
  padding: 1em;
  line-height: 20px;
}

tt, code, pre {
   font-family: 'DejaVu Sans Mono', 'Droid Sans Mono', 'Lucida Console', Consolas, Monaco, monospace;
}

h1 {
   font-size:2.2em;
}

h2 {
   font-size:1.8em;
}

h3 {
   font-size:1.4em;
}

h4 {
   font-size:1.0em;
}

h5 {
   font-size:0.9em;
}

h6 {
   font-size:0.8em;
}

a:visited {
   color: rgb(50%, 0%, 50%);
}

pre, img {
  max-width: 100%;
}
pre {
  overflow-x: auto;
}
pre code {
   display: block; padding: 0.5em;
}

code {
  font-size: 92%;
  border: 1px solid #ccc;
}

code[class] {
  background-color: #F8F8F8;
}

table, td, th {
  border: none;
}

blockquote {
   color:#666666;
   margin:0;
   padding-left: 1em;
   border-left: 0.5em #EEE solid;
}

hr {
   height: 0px;
   border-bottom: none;
   border-top-width: thin;
   border-top-style: dotted;
   border-top-color: #999999;
}

@media print {
   * {
      background: transparent !important;
      color: black !important;
      filter:none !important;
      -ms-filter: none !important;
   }

   body {
      font-size:12pt;
      max-width:100%;
   }

   a, a:visited {
      text-decoration: underline;
   }

   hr {
      visibility: hidden;
      page-break-before: always;
   }

   pre, blockquote {
      padding-right: 1em;
      page-break-inside: avoid;
   }

   tr, img {
      page-break-inside: avoid;
   }

   img {
      max-width: 100% !important;
   }

   @page :left {
      margin: 15mm 20mm 15mm 10mm;
   }

   @page :right {
      margin: 15mm 10mm 15mm 20mm;
   }

   p, h2, h3 {
      orphans: 3; widows: 3;
   }

   h2, h3 {
      page-break-after: avoid;
   }
}
</style>



</head>

<body>
<p>\vspace{5pt}
\footnotesize</p>

<p>\normalsize</p>

<p>\clearpage</p>

<h1>Introduction</h1>

<p>The PharmacoSet class is structured to contain several different types of 
genomic data, as well as drug dose response data. It is built upon base R data 
types and the ExpressionSet class. However, a PharmacoSet object requires the 
presence of specific annotations for the data to be able to function correctly 
with the function provided in PharmacoGx. The structure allows users to then 
interrogate the data on the basis of cells and drugs, removing the need for the 
user to deal with single experiments. We do not recommend creating PharmacoSet 
objects by the end users, rather we recommend contacting us to have the objects 
created for your dataset of interest. However, for completeness and power users 
the process and structure is documented below.</p>

<h1>PharmacoSet Structure</h1>

<p>The base PharmacoSet structure is as follows:</p>

<p><code>annotation</code>:</p>

<ul>
<li><code>name</code>: Acronym of the pharmacogenomic dataset.</li>
<li><code>dateCreated</code>: When the object was created.</li>
</ul>

<p><code>sessionInfo</code>: Software environment used to create the object.</p>

<p><code>call</code>: Set of parameters used to create the object.</p>

<p><code>dataType</code>: Either &#39;sensitivity&#39;, &#39;perturbation&#39;, or &#39;both&#39;</p>

<p><code>cell</code>: data frame annotating all cell lines investigated in the study.</p>

<p><code>drug</code>: data frame annotating all the drugs investigated in the study.</p>

<p><code>molecularProfiles</code>: List of <code>SummarizedExperiment</code> objects containing the 
molecular profiles of the cell lines, such as mutations, gene expressions, or 
copy number variations.</p>

<p><code>sensitivity</code>: </p>

<ul>
<li><code>n</code>: Number of experiments for each cell line treated with a given drug.</li>
<li><code>info</code>: Metadata for each pharmacological experiment.</li>
<li><code>raw</code>: All cell viability measurements at each drug concentration from the 
drug dose-response curves.</li>
<li><code>profiles</code>: Drug sensitivity values summarizing each dose-response curve 
(IC\(_{50}\), AUC, etc.)</li>
</ul>

<p><code>perturbation</code>:</p>

<ul>
<li><code>n</code>: Number of experiments for each cell line perturbed by a given drug, for 
each molecular data type</li>
<li><code>profiles</code>: The metadata for the perturbation experiments is available for 
each molecular type by calling the appropriate info function</li>
</ul>

<p><code>curation</code>: list of data frames containing the mapping between original drug, 
cell line, and tissue names to standardized identifiers.</p>

<h2><code>annotation</code></h2>

<p>The <code>annotation</code> slot contains information pertaining to the PharmacoSet as a 
whole, including the full information on what packages were loaded into R when 
the PharmacoSet was created, as well as the date and call to the constructor. 
The only data that must be entered upon creation of the object is the name of 
the PharmacoSet, which is passed to the constructor.</p>

<h2><code>datasetType</code></h2>

<p>The <code>datasetType</code> slot is a character string which signifies whether the 
PharmacoSet contains drug dose <code>sensitivity</code> data, genomic <code>perturbation</code> data, 
or <code>both</code>. This slot is necessary for the subsetting and intersecting function 
to be aware of what data types they should be looking for.</p>

<h2><code>cell</code></h2>

<p>This slot contains a <code>data.frame</code> which contains information about all the cells 
profiles across all the data types in the PharamcoSet, including both 
perturbation and sensitivity experiments. It is crucial for the rownames of each 
entry in the data frame to be the <strong>unique cell identifier</strong> used across all the 
datatypes for each cell type. The content of this data frame will vary based on 
what information each dataset provides. One of the columns in this data frame 
must be <code>tissueid</code>, predictably.</p>

<h2><code>drug</code></h2>

<p>This slot contains a <code>data.frame</code> which contains information about all the drugs 
profiled across all the data types in the PharmacoSet, including both 
perturbation and sensitivity experiments. Similar to the <code>cell</code> slot, it is 
crucial for the rownames of each entry in the data frame to be the 
<strong>unique compound identifier</strong> used across all the datatypes for each compound. 
Once again, the content in this data frame will vary based on information 
provided by each dataset. </p>

<h2><code>molecularProfiles</code></h2>

<p>The molecular profiles slot of a PharmacoSet object contains all the molecular 
expression data profiled in the dataset. Each type of data is kept in a separate 
SummarizedExperiment (SE) object, and they are all stored as a list. However, to
insure coordination between data types, there are specific annotations that must
be included in the SummarizedExperiment object. The annotations also differ 
slightly in the cases of <code>perturbation</code> and <code>sensitivity</code> datasets. </p>

<p>First of all, each SummarizedExperiment object must be labelled with the type of 
dataincluded within the object. Currently only <code>mutation</code>, <code>fusion</code>, and <code>rna</code> 
molecular data types are recognized and handled differently by summarizing and 
signature generating functions. This labelling is done by calling, for example:</p>

<pre><code class="r">SE@annotation &lt;- &quot;rna&quot;
</code></pre>

<p>The phenoData in each SummarizedExperiment object also needs to contain specific 
columns labelling each experiment. For sensitivity type datasets, this has to 
include:</p>

<p><code>cellid</code>: This column contains a cell identifier that matches <strong>exactly</strong> the 
rownames of the cell slot of the PSet. 
<code>batchid</code>: This column contains and id of the batch of each experiment. If 
such data is not available then it is customary to fill with NA, but this column 
must be present.</p>

<p>Additionally, for the perturbation type datasets, there are several more columns 
of metadata that must be included to enable the modelling of differential gene 
expression. These are:</p>

<p><code>drugid</code>: The identifier of the drug used in each experiment when a compound 
was applied. For controls, it should be left as NA. These must match <strong>exactly</strong> 
to the rownames of the drug slot in the PharmacoSet. </p>

<p><code>duration</code>: If available, the ength of the experiment. Should be 0 for 
controls. </p>

<p><code>concentration</code>: For experiments where compounds were applied, the 
concentration. Should be 0 for controls. </p>

<p><code>xptype</code>: A label of either <code>perturbation</code>, of <code>control</code> respectively.</p>

<p>Otherwise each SummarizedExperiment object should be constructed as specified 
by the SummarizedExperiment package. </p>

<p>Once these SummarizedExperiment objects are created, they should be added to a 
list, with the name of each object in the list being descriptive of the type of 
data contained. </p>

<h2><code>sensitivity</code></h2>

<p>This is a list containing all the data pertaining to drug dose response 
experiments, usually only present in <code>sensitivity</code> or <code>both</code> type datasets. This
includes the names the following names slots in the list:</p>

<p><code>info</code>: Metadata for each pharmacological experiment stored in a <code>data.frame</code>. 
Each row of this data.frame should be labelled with a unique experiment 
identifier. This data.frame must also contain specific columns: </p>

<p><code>cellid</code>: Contains a cell identifier that matches <strong>exactly</strong> the rownames of 
the cell slot of the PSet.</p>

<p><code>drugid</code>: The identifier of the drug used in each experiment when a compound was 
applied. These must match <strong>exactly</strong> to the rownames of the drug slot in the 
PharmacoSet.</p>

<p><code>raw</code>: This is a 3-D <code>array</code> of raw drug dose response data. The first 
dimension is the experiments, with the names of each row in this dimension 
labelled exactly as the experiment ids used in the metadata data.frame <code>info</code> 
above. The second dimension is the doses, with each &ldquo;column&rdquo; labelled as 
<code>doses</code>, where the number is the column number. This is as long as there are 
different number of doses for each experiment. The third dimension is always 
fixed at 2, with names <code>Dose</code>, and <code>Viability</code>. Dose} contains the actual dose 
used (usually in micro-molar), while viability contains the cell viability 
measurement at that dose. Therefore, for each experiment there are two vectors, 
one of doses and one of viabilities at each dose.</p>

<p><code>profiles</code> This is a data.frame containing down the rows, with <strong>exactly</strong> 
matching rownames, the experiments as labelled in the <code>info</code> data.frame, and 
each column being a summary of the drug dose sensitivity, such as
<code>auc\_published</code> or <code>ic50\_published</code>, as published with the data.</p>

<p><code>n</code> This is a <code>matrix</code> containing cellids matching exactly the rownames of the 
cell slot of the PSet down the rows, and drugids matching exactly the rownames 
of the drug slot in the columns. This data.frame summaries how many experiments 
are in the data for each pair, allowing quick identification of what pairs were 
tested together. Note that this does not need to be generated before 
constructing the object, it is generated by the constructor.</p>

<h2><code>perturbation</code></h2>

<p>This slot is fully filled by the constructor, so nothing needs to be created for 
it.</p>

<p><code>n</code>: This is a 3-D <code>array</code> containing cellids matching exactly the rownames of 
the cell slot of the PSet down the rows, and drugids matching exactly the 
rownames of the drug slot in the columns, and the third dimension being the 
names of the different molecular profile types. This array summarizes how many 
perturbation experiments are in each molecular data type for each pair, allowing 
quick identification of what pairs were tested together.</p>

<p><code>info</code>: This is always exactly the string referring to each data type 
separately: &#39;The metadata for the perturbation experiments is available for each 
molecular type by calling the appropriate info function&#39;</p>

<h2><code>curation</code></h2>

<p>This slot contains three data.frames, one for drugs, tissues, and cells. Each 
contains two columns, the first with the unique identifier that is used between 
all PharmacoSet objects for this drug, and the second with the identifier used 
within the dataset. At this time there is no set method to find the unique 
identifiers between all PharmacoSets for arbitrary drugs, as much labour 
intensive curation has gone into carefully matching identifiers between datasets. 
However, these tables are used only by the intersecting function, and all other 
functions will function without this data, therefore for PharmacoSet objects 
created by users, we recommend just filling both columns with the identifiers 
used in the datasets, to allow for at least some matching to be done. </p>

<h1>Creating the PharmacoSet</h1>

<p>Once all the data is in the proper format, it can be passed to a the constructor 
function as below:</p>

<pre><code class="r">PharmacoSet(name, 
            molecularProfiles=list(), 
            cell=data.frame(), 
            drug=data.frame(), 
            sensitivityInfo=data.frame(),
            sensitivityRaw=array(dim=c(0,0,0)), 
            sensitivityProfiles=matrix(), 
            curationDrug=data.frame(), 
            curationCell=data.frame(), 
            curationTissue=data.frame(), 
            datasetType=c(&quot;sensitivity&quot;, &quot;perturbation&quot;, &quot;both&quot;),
            verify = TRUE)
</code></pre>

<p>Here, <code>name</code> is the name of the PharmacoSet as described in annotations above, 
<code>molecularProfiles</code> is the list of ExpressionSet objects described above, cell 
and drug are the data.frames in the cell and drug slots. The <code>sensitivityInfo</code>, 
<code>sensitivityRaw</code>, <code>sensitivityProfiles</code> are the elements of the <code>sensitivity</code> 
slot, and curationDrug, curationCell, and curationTissue the contents of the 
curation slot. Finally, <code>datasetType</code> is a string to signify which type of data 
isincluded in the datset, and verify is a flag for running the verification of 
the pSet after its construction. Any data detailed above but missing from this 
constructor call will be created by the constructor, and the default empty 
values are as in the call above. Notice that none of the values should be set 
as NA, they should be ommited if the data is not being provided. </p>

</body>

</html>
